File: typescript/design_pattern/facade/main_facade.ts
/**
 * Facade Design Pattern in TypeScript
 *
 * The Facade pattern provides a simplified interface to a complex subsystem.
 * It hides the complexity of the subsystem and provides a single unified
 * interface that is easier to use for clients.
 */

// ============================================================
// Subsystem Classes - Complex components that the Facade simplifies
// ============================================================

// Subsystem 1: Television
class Television {
    private isOn: boolean = false;
    private channel: number = 1;

    on(): void {
        this.isOn = true;
        console.log("[Line 18] Television: Powering on");
    }

    off(): void {
        this.isOn = false;
        console.log("[Line 23] Television: Powering off");
    }

    setInputChannel(channel: number): void {
        this.channel = channel;
        console.log(`[Line 28] Television: Setting input channel to ${channel}`);
    }
}

// Subsystem 2: Sound System
class SoundSystem {
    private volume: number = 0;
    private surroundMode: boolean = false;

    on(): void {
        console.log("[Line 37] SoundSystem: Powering on amplifier");
    }

    off(): void {
        console.log("[Line 41] SoundSystem: Powering off amplifier");
    }

    setVolume(level: number): void {
        this.volume = level;
        console.log(`[Line 46] SoundSystem: Setting volume to ${level}`);
    }

    setSurroundSound(enabled: boolean): void {
        this.surroundMode = enabled;
        const mode = enabled ? "surround" : "stereo";
        console.log(`[Line 52] SoundSystem: Setting audio mode to ${mode}`);
    }
}

// Subsystem 3: Streaming Player
class StreamingPlayer {
    private currentMovie: string = "";

    on(): void {
        console.log("[Line 60] StreamingPlayer: Starting streaming device");
    }

    off(): void {
        console.log("[Line 64] StreamingPlayer: Shutting down streaming device");
    }

    play(movie: string): void {
        this.currentMovie = movie;
        console.log(`[Line 69] StreamingPlayer: Playing "${movie}"`);
    }

    pause(): void {
        console.log(`[Line 73] StreamingPlayer: Pausing "${this.currentMovie}"`);
    }

    stop(): void {
        console.log(`[Line 77] StreamingPlayer: Stopping "${this.currentMovie}"`);
        this.currentMovie = "";
    }
}

// Subsystem 4: Room Lights
class RoomLights {
    private brightness: number = 100;

    on(): void {
        this.brightness = 100;
        console.log("[Line 87] RoomLights: Turning lights on");
    }

    off(): void {
        this.brightness = 0;
        console.log("[Line 92] RoomLights: Turning lights off");
    }

    dim(level: number): void {
        this.brightness = level;
        console.log(`[Line 97] RoomLights: Dimming lights to ${level}%`);
    }
}

// Subsystem 5: Popcorn Machine
class PopcornMachine {
    on(): void {
        console.log("[Line 104] PopcornMachine: Turning on popcorn machine");
    }

    off(): void {
        console.log("[Line 108] PopcornMachine: Turning off popcorn machine");
    }

    pop(): void {
        console.log("[Line 112] PopcornMachine: Popping popcorn!");
    }
}

// ============================================================
// Facade Class - Provides simplified interface to the subsystems
// ============================================================

class HomeTheaterFacade {
    private tv: Television;
    private soundSystem: SoundSystem;
    private streamingPlayer: StreamingPlayer;
    private lights: RoomLights;
    private popcorn: PopcornMachine;

    constructor(
        tv: Television,
        soundSystem: SoundSystem,
        streamingPlayer: StreamingPlayer,
        lights: RoomLights,
        popcorn: PopcornMachine
    ) {
        this.tv = tv;
        this.soundSystem = soundSystem;
        this.streamingPlayer = streamingPlayer;
        this.lights = lights;
        this.popcorn = popcorn;
        console.log("[Line 136] HomeTheaterFacade: Home theater system initialized");
    }

    // Simplified method to start watching a movie
    watchMovie(movie: string): void {
        console.log(`\n[Line 141] HomeTheaterFacade: Preparing to watch "${movie}"...`);
        console.log("----------------------------------------");

        this.popcorn.on();
        this.popcorn.pop();
        this.lights.dim(10);
        this.tv.on();
        this.tv.setInputChannel(3);
        this.soundSystem.on();
        this.soundSystem.setVolume(25);
        this.soundSystem.setSurroundSound(true);
        this.streamingPlayer.on();
        this.streamingPlayer.play(movie);

        console.log("----------------------------------------");
        console.log(`[Line 156] HomeTheaterFacade: Enjoy your movie!`);
    }

    // Simplified method to pause the movie
    pauseMovie(): void {
        console.log("\n[Line 161] HomeTheaterFacade: Pausing movie...");
        this.streamingPlayer.pause();
        this.lights.dim(50);
    }

    // Simplified method to end movie watching
    endMovie(): void {
        console.log("\n[Line 168] HomeTheaterFacade: Shutting down home theater...");
        console.log("----------------------------------------");

        this.streamingPlayer.stop();
        this.streamingPlayer.off();
        this.soundSystem.off();
        this.tv.off();
        this.lights.on();
        this.popcorn.off();

        console.log("----------------------------------------");
        console.log("[Line 179] HomeTheaterFacade: Home theater shutdown complete");
    }

    // Simplified method for listening to music
    listenToMusic(): void {
        console.log("\n[Line 184] HomeTheaterFacade: Setting up music mode...");
        console.log("----------------------------------------");

        this.lights.dim(60);
        this.soundSystem.on();
        this.soundSystem.setVolume(15);
        this.soundSystem.setSurroundSound(false);

        console.log("----------------------------------------");
        console.log("[Line 193] HomeTheaterFacade: Music mode ready");
    }

    // End music mode
    endMusic(): void {
        console.log("\n[Line 198] HomeTheaterFacade: Ending music mode...");
        this.soundSystem.off();
        this.lights.on();
        console.log("[Line 201] HomeTheaterFacade: Music mode ended");
    }
}

// ============================================================
// Second Example: Computer Startup Facade
// ============================================================

// Subsystem: CPU
class CPU {
    freeze(): void {
        console.log("[Line 212] CPU: Freezing processor");
    }

    jump(address: number): void {
        console.log(`[Line 216] CPU: Jumping to address ${address}`);
    }

    execute(): void {
        console.log("[Line 220] CPU: Executing instructions");
    }
}

// Subsystem: Memory
class Memory {
    load(address: number, data: string): void {
        console.log(`[Line 227] Memory: Loading "${data}" at address ${address}`);
    }
}

// Subsystem: HardDrive
class HardDrive {
    read(sector: number, size: number): string {
        const data = `BootData[sector:${sector},size:${size}]`;
        console.log(`[Line 235] HardDrive: Reading ${size} bytes from sector ${sector}`);
        return data;
    }
}

// Computer Facade
class ComputerFacade {
    private cpu: CPU;
    private memory: Memory;
    private hardDrive: HardDrive;

    constructor() {
        this.cpu = new CPU();
        this.memory = new Memory();
        this.hardDrive = new HardDrive();
        console.log("[Line 250] ComputerFacade: Computer components initialized");
    }

    start(): void {
        console.log("\n[Line 254] ComputerFacade: Starting computer...");
        console.log("----------------------------------------");

        this.cpu.freeze();
        const bootData = this.hardDrive.read(0, 1024);
        this.memory.load(0x00, bootData);
        this.cpu.jump(0x00);
        this.cpu.execute();

        console.log("----------------------------------------");
        console.log("[Line 264] ComputerFacade: Computer started successfully");
    }

    shutdown(): void {
        console.log("\n[Line 268] ComputerFacade: Shutting down computer...");
        console.log("[Line 269] ComputerFacade: Saving state and powering off");
    }
}

// ============================================================
// Demonstration
// ============================================================

function main(): void {
    console.log("=== Facade Pattern Demonstration ===");

    // Demo 1: Home Theater Facade
    console.log("\n--- Home Theater Facade Demo ---");

    // Create subsystem components
    const tv = new Television();
    const soundSystem = new SoundSystem();
    const streamingPlayer = new StreamingPlayer();
    const lights = new RoomLights();
    const popcorn = new PopcornMachine();

    // Create facade with all subsystems
    const homeTheater = new HomeTheaterFacade(
        tv,
        soundSystem,
        streamingPlayer,
        lights,
        popcorn
    );

    // Use simplified facade methods instead of controlling each subsystem
    homeTheater.watchMovie("The Matrix");

    // Pause for a break
    homeTheater.pauseMovie();

    // End the movie session
    homeTheater.endMovie();

    // Quick music session
    homeTheater.listenToMusic();
    homeTheater.endMusic();

    // Demo 2: Computer Startup Facade
    console.log("\n\n--- Computer Startup Facade Demo ---");

    const computer = new ComputerFacade();
    computer.start();
    computer.shutdown();

    // Demo 3: Show what it would look like WITHOUT the facade
    console.log("\n\n--- Without Facade (Direct Subsystem Control) ---");
    console.log("[Line 318] Direct control requires managing each subsystem:");

    const manualTv = new Television();
    const manualSound = new SoundSystem();
    const manualPlayer = new StreamingPlayer();
    const manualLights = new RoomLights();
    const manualPopcorn = new PopcornMachine();

    console.log("\nManually setting up movie (verbose and error-prone):");
    manualPopcorn.on();
    manualPopcorn.pop();
    manualLights.dim(10);
    manualTv.on();
    manualTv.setInputChannel(3);
    manualSound.on();
    manualSound.setVolume(25);
    manualSound.setSurroundSound(true);
    manualPlayer.on();
    manualPlayer.play("Inception");

    console.log("\n[Line 338] Notice how the facade simplifies all of this!");
    console.log("         With facade: homeTheater.watchMovie('Inception')");
    console.log("         Without facade: 10+ individual method calls");

    console.log("\n=== End of Demonstration ===");
}

main();

File: rust/design_pattern/facade/main_facade.rs
// Facade Design Pattern in Rust
// Demonstrates how a facade simplifies interaction with complex subsystems

// ============================================================================
// Subsystem Components - Complex internal systems
// ============================================================================

/// Television subsystem with multiple operations
struct Television {
    brand: String,
}

impl Television {
    fn new(brand: &str) -> Self {
        Television {
            brand: brand.to_string(),
        }
    }

    fn turn_on(&self) {
        println!("  [TV] {} Television is turning ON", self.brand);
    }

    fn turn_off(&self) {
        println!("  [TV] {} Television is turning OFF", self.brand);
    }

    fn set_input(&self, input: &str) {
        println!("  [TV] Setting input source to: {}", input);
    }

    fn set_brightness(&self, level: u8) {
        println!("  [TV] Adjusting brightness to {}%", level);
    }
}

/// Sound system subsystem with audio controls
struct SoundSystem {
    model: String,
}

impl SoundSystem {
    fn new(model: &str) -> Self {
        SoundSystem {
            model: model.to_string(),
        }
    }

    fn turn_on(&self) {
        println!("  [Audio] {} Sound System powering ON", self.model);
    }

    fn turn_off(&self) {
        println!("  [Audio] {} Sound System powering OFF", self.model);
    }

    fn set_volume(&self, level: u8) {
        println!("  [Audio] Setting volume to {}", level);
    }

    fn set_surround_mode(&self, mode: &str) {
        println!("  [Audio] Surround mode set to: {}", mode);
    }
}

/// Streaming player subsystem
struct StreamingPlayer {
    service: String,
}

impl StreamingPlayer {
    fn new(service: &str) -> Self {
        StreamingPlayer {
            service: service.to_string(),
        }
    }

    fn turn_on(&self) {
        println!("  [Stream] {} Streaming Player starting", self.service);
    }

    fn turn_off(&self) {
        println!("  [Stream] {} Streaming Player shutting down", self.service);
    }

    fn login(&self) {
        println!("  [Stream] Logging into {} account", self.service);
    }

    fn play(&self, content: &str) {
        println!("  [Stream] Now playing: {}", content);
    }

    fn stop(&self) {
        println!("  [Stream] Playback stopped");
    }
}

/// Room lighting subsystem
struct Lights {
    room: String,
}

impl Lights {
    fn new(room: &str) -> Self {
        Lights {
            room: room.to_string(),
        }
    }

    fn dim(&self, level: u8) {
        println!("  [Lights] {} lights dimmed to {}%", self.room, level);
    }

    fn on(&self) {
        println!("  [Lights] {} lights turned ON (100%)", self.room);
    }
}

// ============================================================================
// Facade - Simplified interface to the complex subsystems
// ============================================================================

/// HomeTheaterFacade provides a simple interface to control
/// all home theater components with single method calls
struct HomeTheaterFacade {
    tv: Television,
    sound: SoundSystem,
    player: StreamingPlayer,
    lights: Lights,
}

impl HomeTheaterFacade {
    fn new() -> Self {
        println!("=== Initializing Home Theater System ===\n");
        HomeTheaterFacade {
            tv: Television::new("Samsung"),
            sound: SoundSystem::new("Sonos"),
            player: StreamingPlayer::new("Netflix"),
            lights: Lights::new("Living Room"),
        }
    }

    /// Simple method to start movie night - hides all complexity
    fn watch_movie(&self, movie: &str) {
        println!(">>> Starting Movie Night: '{}' <<<\n", movie);

        // Facade coordinates all subsystems
        self.lights.dim(20);
        self.tv.turn_on();
        self.tv.set_input("HDMI-1");
        self.tv.set_brightness(80);
        self.sound.turn_on();
        self.sound.set_volume(50);
        self.sound.set_surround_mode("Cinema");
        self.player.turn_on();
        self.player.login();
        self.player.play(movie);

        println!("\n>>> Enjoy your movie! <<<\n");
    }

    /// Simple method to end movie night - hides shutdown complexity
    fn end_movie(&self) {
        println!(">>> Ending Movie Night <<<\n");

        // Facade coordinates orderly shutdown
        self.player.stop();
        self.player.turn_off();
        self.sound.turn_off();
        self.tv.turn_off();
        self.lights.on();

        println!("\n>>> System shutdown complete <<<\n");
    }

    /// Simple method for music listening mode
    fn listen_to_music(&self) {
        println!(">>> Starting Music Mode <<<\n");

        self.lights.dim(60);
        self.sound.turn_on();
        self.sound.set_volume(40);
        self.sound.set_surround_mode("Stereo");

        println!("\n>>> Music mode ready <<<\n");
    }
}

// ============================================================================
// Main - Demonstrates facade usage
// ============================================================================

fn main() {
    println!("============================================");
    println!("   FACADE DESIGN PATTERN DEMONSTRATION");
    println!("============================================\n");

    // Create the facade - user only interacts with this
    let home_theater = HomeTheaterFacade::new();

    // Without facade: user would need to call 10+ methods manually
    // With facade: single method call handles everything

    println!("--------------------------------------------");
    println!("Client calls: home_theater.watch_movie(...)");
    println!("--------------------------------------------\n");

    home_theater.watch_movie("The Matrix");

    println!("--------------------------------------------");
    println!("Client calls: home_theater.end_movie()");
    println!("--------------------------------------------\n");

    home_theater.end_movie();

    println!("--------------------------------------------");
    println!("Client calls: home_theater.listen_to_music()");
    println!("--------------------------------------------\n");

    home_theater.listen_to_music();

    println!("============================================");
    println!("   KEY INSIGHT: Facade Pattern");
    println!("============================================");
    println!("The client made only 3 simple calls, but the");
    println!("facade coordinated 15+ subsystem operations!");
    println!("============================================\n");
}
